tutorials/scheduler.ipynb

```diff
@@ -0,0 +1,1043 @@
+    "\n",
+    "The following runner and metric set up intractions between the `Scheduler` and the mock external system we assume:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ax.core.runner import Runner\n",
+    "from ax.core.base_trial import BaseTrial\n",
+    "from ax.core.trial import Trial\n",
+    "\n",
+    "class MockJobRunner(Runner):  # Deploys trials to external system.\n",
+    "    \n",
+    "    def run(self, trial: BaseTrial) -> Dict[str,Any]:\n",
+    "        \"\"\"Deploys a trial and returns dict of run metadata.\"\"\"\n",
+    "        if not isinstance(trial, Trial):\n",
+    "            raise ValueError(\"This runner only handles `Trial`.\")\n",
+    "        \n",
+    "        mock_job_queue = get_mock_job_queue_client()\n",
+    "        job_id = mock_job_queue.schedule_job_with_parameters(\n",
+    "            parameters=trial.arm.parameters\n",
+    "        )\n",
+    "        # This run metadata will be attached to trial as `trial.run_metadata`\n",
+    "        # by the base `Scheduler`.\n",
+    "        return {\"job_id\": job_id}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "\n",
+    "from ax.core.metric import Metric\n",
+    "from ax.core.base_trial import BaseTrial\n",
+    "from ax.core.data import Data\n",
+    "\n",
+    "class BraninForMockJobMetric(Metric):  # Pulls data for trial from external system.\n",
+    "    \n",
+    "    def fetch_trial_data(self, trial: BaseTrial) -> Data:\n",
+    "        \"\"\"Obtains data via fetching it from ` for a given trial.\"\"\"\n",
+    "        if not isinstance(trial, Trial):\n",
+    "            raise ValueError(\"This metric only handles `Trial`.\")\n",
+    "        \n",
+    "        mock_job_queue = get_mock_job_queue_client()\n",
+    "        \n",
+    "        # Here we leverage the \"job_id\" metadata created by `MockJobRunner.run`.\n",
+    "        branin_data = mock_job_queue.get_outcome_value_for_completed_job(\n",
+    "            job_id=trial.run_metadata.get(\"job_id\")\n",
+    "        )\n",
+    "        df_dict = {\n",
+    "            \"trial_index\": trial.index,\n",
+    "            \"metric_name\": \"branin\",\n",
+    "            \"arm_name\": trial.arm.name,\n",
+    "            \"mean\": branin_data.get(\"branin\"),\n",
+    "            # Can be set to 0.0 if function is known to be noiseless\n",
+    "            # or to an actual value when SEM is known. Setting SEM to\n",
+    "            # `None` results in Ax assuming unknown noise and inferring\n",
+    "            # noise level from data.\n",
+    "            \"sem\": None,\n",
+    "        }\n",
+    "        return Data(df=pd.DataFrame.from_records([df_dict]))"
+   ]
+  },
+  {
+   "cell_type"
```

Comment:
```suggestion
    "        status_dict = defaultdict(set)\n",
```

Reply:
ugh somehow the json messes up the code suggestion. You know what I mean, I hope.

Reply:
Yeah for sure! I was a bit worried that not everyone is familiar with `defaultdict` and felt like the overhead here was low enough to not use it, but I think you're right that anyone dealing with this tutorial is probably familiar with it : )

Reply:
If they aren't yet they should be, so teaching them is an added benefit of the tutorial :)